File: desktop/macos/Sources/VoiceBridgeAI/Session/TranslationRecorder.swift
import AppKit
import Foundation

/// Persists finalized bilingual subtitle lines to transcript files.
@MainActor
final class TranslationRecorder {
    static let shared = TranslationRecorder()

    private struct Entry {
        let segmentId: String
        var english: String
        var chinese: String
        var revised: Bool
        let firstSeen: Date
        var lastUpdated: Date
    }

    private var entries: [String: Entry] = [:]
    private var sessionURL: URL?
    private var sessionStarted: Date?
    private var sessionEnded: String?
    private var reviseModeId: String = ""
    private var reviseModeLabel: String = ""

    private init() {}

    var transcriptsDirectory: URL {
        TranscriptPreferences.storageDirectory
    }

    var currentSessionURL: URL? { sessionURL }

    func beginSession(reviseModeId: String, reviseModeLabel: String) {
        guard TranscriptPreferences.recordEnabled else { return }
        entries.removeAll()
        sessionStarted = Date()
        sessionEnded = nil
        self.reviseModeId = reviseModeId
        self.reviseModeLabel = reviseModeLabel

        let name = TranscriptPreferences.sessionFilename(started: sessionStarted!, modeId: reviseModeId)
        sessionURL = transcriptsDirectory.appendingPathComponent(name)
        flushToDisk(ended: false)
    }

    func beginSessionIfNeeded(reviseModeId: String, reviseModeLabel: String) {
        guard sessionURL == nil else { return }
        beginSession(reviseModeId: reviseModeId, reviseModeLabel: reviseModeLabel)
    }

    func record(segmentId: String, english: String, chinese: String, revised: Bool) {
        guard TranscriptPreferences.recordEnabled else { return }
        let en = english.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !en.isEmpty else { return }

        if sessionURL == nil {
            let health = SettingsStore.shared.health
            let modeId = SettingsStore.shared.engine.reviseMode
            let label = ReviseModeGuides.info(for: modeId, health: health)?.label ?? modeId
            beginSession(reviseModeId: modeId, reviseModeLabel: label)
        }

        let now = Date()
        if var existing = entries[segmentId] {
            existing.english = en
            existing.chinese = chinese.trimmingCharacters(in: .whitespacesAndNewlines)
            existing.revised = existing.revised || revised
            existing.lastUpdated = now
            entries[segmentId] = existing
        } else {
            entries[segmentId] = Entry(
                segmentId: segmentId,
                english: en,
                chinese: chinese.trimmingCharacters(in: .whitespacesAndNewlines),
                revised: revised,
                firstSeen: now,
                lastUpdated: now
            )
        }
        flushToDisk(ended: false)
    }

    func endSession() {
        guard sessionURL != nil else { return }
        sessionEnded = Self.displayFormatter.string(from: Date())
        flushToDisk(ended: true)
        entries.removeAll()
        sessionURL = nil
        sessionStarted = nil
        sessionEnded = nil
    }

    func openTranscriptsDirectory() {
        do {
            try FileManager.default.createDirectory(at: transcriptsDirectory, withIntermediateDirectories: true)
            NSWorkspace.shared.open(transcriptsDirectory)
        } catch {
            NSLog("TranslationRecorder: failed to open transcripts dir: \(error.localizedDescription)")
        }
    }

    static func convertFile(
        at source: URL,
        layout: TranscriptContentLayout,
        format: TranscriptPreferences.FileFormat
    ) throws -> URL {
        let document = try TranscriptParser.parse(file: source)
        if document.entries.isEmpty {
            throw TranscriptConvertError.noEntries
        }
        let body = TranscriptRenderer.render(
            document: document,
            layout: layout,
            format: format,
            ended: document.ended != nil
        )
        let dest = TranscriptRenderer.exportURL(for: source, layout: layout, format: format)
        try FileManager.default.createDirectory(
            at: dest.deletingLastPathComponent(),
            withIntermediateDirectories: true
        )
        try body.write(to: dest, atomically: true, encoding: .utf8)
        return dest
    }

    private func flushToDisk(ended: Bool) {
        guard let url = sessionURL else { return }
        do {
            try FileManager.default.createDirectory(at: transcriptsDirectory, withIntermediateDirectories: true)
            let body = TranscriptRenderer.render(
                document: buildDocument(ended: ended),
                layout: TranscriptPreferences.contentLayout,
                format: TranscriptPreferences.fileFormat,
                ended: ended
            )
            try body.write(to: url, atomically: true, encoding: .utf8)
        } catch {
            NSLog("TranslationRecorder: write failed: \(error.localizedDescription)")
        }
    }

    private func buildDocument(ended: Bool) -> TranscriptDocument {
        let started = sessionStarted.map { Self.displayFormatter.string(from: $0) }
        let scene = reviseModeLabel.isEmpty
            ? reviseModeId
            : "\(reviseModeLabel)（\(reviseModeId)）"

        let ordered = entries.values.sorted {
            (Int($0.segmentId) ?? 0) < (Int($1.segmentId) ?? 0)
        }

        let transcriptEntries = ordered.enumerated().map { offset, entry in
            TranscriptEntry(
                index: offset + 1,
                time: Self.displayFormatter.string(from: entry.lastUpdated),
                revised: entry.revised,
                english: entry.english,
                chinese: entry.chinese
            )
        }

        return TranscriptDocument(
            started: started,
            scene: scene,
            ended: ended ? (sessionEnded ?? Self.displayFormatter.string(from: Date())) : nil,
            entries: transcriptEntries
        )
    }

    private static let displayFormatter: DateFormatter = {
        let f = DateFormatter()
        f.locale = Locale(identifier: "zh_CN")
        f.dateFormat = "yyyy-MM-dd HH:mm:ss"
        return f
    }()
}

enum TranscriptConvertError: LocalizedError {
    case noEntries

    var errorDescription: String? {
        switch self {
        case .noEntries: return "未能从文件中解析出字幕条目"
        }
    }
}

File: desktop/windows/VoiceBridgeAI/VoiceBridgeAI/Session/TranslationRecorder.cs
using System.Globalization;
using VoiceBridgeAI.Settings;

namespace VoiceBridgeAI.Session;

public sealed class TranslationRecorder
{
    public static TranslationRecorder Shared { get; } = new();

    private sealed class Entry
    {
        public required string SegmentId { get; init; }
        public string English { get; set; } = "";
        public string Chinese { get; set; } = "";
        public bool Revised { get; set; }
        public DateTime FirstSeen { get; init; }
        public DateTime LastUpdated { get; set; }
    }

    private readonly Dictionary<string, Entry> _entries = new();
    private string? _sessionPath;
    private DateTime? _sessionStarted;
    private string? _sessionEnded;
    private string _reviseModeId = "";
    private string _reviseModeLabel = "";

    private TranslationRecorder()
    {
    }

    public string TranscriptsDirectory => TranscriptPreferences.StorageDirectory;

    public string? CurrentSessionPath => _sessionPath;

    public void BeginSession(string reviseModeId, string reviseModeLabel)
    {
        if (!TranscriptPreferences.RecordEnabled)
        {
            return;
        }

        _entries.Clear();
        _sessionStarted = DateTime.Now;
        _sessionEnded = null;
        _reviseModeId = reviseModeId;
        _reviseModeLabel = reviseModeLabel;

        var name = TranscriptPreferences.SessionFilename(_sessionStarted.Value, reviseModeId);
        _sessionPath = Path.Combine(TranscriptsDirectory, name);
        FlushToDisk(ended: false);
    }

    public void BeginSessionIfNeeded(string reviseModeId, string reviseModeLabel)
    {
        if (_sessionPath is null)
        {
            BeginSession(reviseModeId, reviseModeLabel);
        }
    }

    public void Record(string segmentId, string english, string chinese, bool revised)
    {
        if (!TranscriptPreferences.RecordEnabled)
        {
            return;
        }

        var en = english.Trim();
        if (string.IsNullOrEmpty(en))
        {
            return;
        }

        if (_sessionPath is null)
        {
            try
            {
                var modeId = SettingsStore.Shared.Engine.ReviseMode;
                var label = ReviseModeLabel(modeId);
                BeginSession(modeId, label);
            }
            catch
            {
                BeginSession("speech", "speech");
            }
        }

        var now = DateTime.Now;
        if (_entries.TryGetValue(segmentId, out var existing))
        {
            existing.English = en;
            existing.Chinese = chinese.Trim();
            existing.Revised = existing.Revised || revised;
            existing.LastUpdated = now;
        }
        else
        {
            _entries[segmentId] = new Entry
            {
                SegmentId = segmentId,
                English = en,
                Chinese = chinese.Trim(),
                Revised = revised,
                FirstSeen = now,
                LastUpdated = now,
            };
        }

        FlushToDisk(ended: false);
    }

    public void EndSession()
    {
        if (_sessionPath is null)
        {
            return;
        }

        _sessionEnded = DisplayFormatter.Format(DateTime.Now);
        FlushToDisk(ended: true);
        _entries.Clear();
        _sessionPath = null;
        _sessionStarted = null;
        _sessionEnded = null;
    }

    public void OpenTranscriptsDirectory()
    {
        try
        {
            Directory.CreateDirectory(TranscriptsDirectory);
            System.Diagnostics.Process.Start(new System.Diagnostics.ProcessStartInfo
            {
                FileName = TranscriptsDirectory,
                UseShellExecute = true,
            });
        }
        catch
        {
            // Opening Explorer failure should not crash the app.
        }
    }

    public static string ConvertFile(
        string sourcePath,
        TranscriptPreferences.ContentLayout layout,
        TranscriptPreferences.FileFormat format)
    {
        var document = TranscriptParser.ParseFile(sourcePath);
        if (document.Entries.Count == 0)
        {
            throw new TranscriptConvertException("未能从文件中解析出字幕条目");
        }

        var body = TranscriptRenderer.Render(
            document,
            layout,
            format,
            ended: !string.IsNullOrWhiteSpace(document.Ended));
        var dest = TranscriptRenderer.ExportPath(sourcePath, layout, format);
        Directory.CreateDirectory(Path.GetDirectoryName(dest)!);
        File.WriteAllText(dest, body, System.Text.Encoding.UTF8);
        return dest;
    }

    private void FlushToDisk(bool ended)
    {
        if (_sessionPath is null)
        {
            return;
        }

        try
        {
            Directory.CreateDirectory(TranscriptsDirectory);
            var body = TranscriptRenderer.Render(
                BuildDocument(ended),
                TranscriptPreferences.ContentLayoutValue,
                TranscriptPreferences.FileFormatValue,
                ended);
            File.WriteAllText(_sessionPath, body, System.Text.Encoding.UTF8);
        }
        catch
        {
            // Write failure should not crash the session.
        }
    }

    private TranscriptDocument BuildDocument(bool ended)
    {
        var started = _sessionStarted.HasValue
            ? DisplayFormatter.Format(_sessionStarted.Value)
            : null;
        var scene = string.IsNullOrEmpty(_reviseModeLabel)
            ? _reviseModeId
            : $"{_reviseModeLabel}（{_reviseModeId}）";

        var ordered = _entries.Values
            .OrderBy(e => int.TryParse(e.SegmentId, out var n) ? n : 0)
            .ToList();

        var transcriptEntries = ordered.Select((entry, offset) => new TranscriptEntry
        {
            Index = offset + 1,
            Time = DisplayFormatter.Format(entry.LastUpdated),
            Revised = entry.Revised,
            English = entry.English,
            Chinese = entry.Chinese,
        }).ToList();

        return new TranscriptDocument
        {
            Started = started,
            Scene = scene,
            Ended = ended ? (_sessionEnded ?? DisplayFormatter.Format(DateTime.Now)) : null,
            Entries = transcriptEntries,
        };
    }

    private static string ReviseModeLabel(string modeId)
    {
        try
        {
            var health = SettingsStore.Shared.Health;
            return ProviderOption.List(health, "reviseModes")
                .FirstOrDefault(p => p.Id == modeId).Label ?? modeId;
        }
        catch
        {
            return modeId;
        }
    }

    private static readonly CultureInfo DisplayCulture = CultureInfo.GetCultureInfo("zh-CN");

    private static class DisplayFormatter
    {
        public static string Format(DateTime value) =>
            value.ToString("yyyy-MM-dd HH:mm:ss", DisplayCulture);
    }
}
